src/protocols/bdx/BdxMessages.cpp

```diff
@@ -0,0 +1,401 @@
+    }
+
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    aBuffer.SetDataLength(static_cast<uint16_t>(bbuf.Written()));
+
+exit:
+    return err;
+}
+
+CHIP_ERROR TransferInit::Parse(const System::PacketBuffer & aBuffer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t proposedTransferCtl;
+    uint8_t rangeCtl;
+    uint32_t tmpUint32Value = 0; // Used for reading non-wide length and offset fields
+    uint8_t * bufStart      = aBuffer.Start();
+    Reader bufReader(bufStart, aBuffer.DataLength());
+    BitFlags<uint8_t, RangeControlFlags> rangeCtlFlags;
+
+    SuccessOrExit(bufReader.Read8(&proposedTransferCtl).Read8(&rangeCtl).Read16(&mMaxBlockSize).StatusCode());
+
+    mSupportedVersions = proposedTransferCtl & kVersionMask;
+    mTransferCtlFlags.SetRaw(static_cast<uint8_t>(proposedTransferCtl & ~kVersionMask));
+    rangeCtlFlags.SetRaw(rangeCtl);
+
+    if (rangeCtlFlags.Has(kStartOffset))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&mStartOffset).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            mStartOffset = tmpUint32Value;
+        }
+    }
+
+    if (rangeCtlFlags.Has(kDefLen))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&mMaxLength).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            mMaxLength = tmpUint32Value;
+        }
+    }
+
+    SuccessOrExit(bufReader.Read16(&mFileDesLength).StatusCode());
+
+    // WARNING: this struct will store a pointer to the start of the file designator in the PacketBuffer,
+    // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+    // or there is risk of unsafe memory access.
+    VerifyOrExit(bufReader.HasAtLeast(mFileDesLength), err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    mFileDesignator = &bufStart[bufReader.OctetsRead()];
+
+    // Rest of message is metadata (could be empty)
+    if (bufReader.Remaining() > mFileDesLength)
+    {
+        // WARNING: this struct will store a pointer to the start of metadata in the PacketBuffer,
+        // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+        // or there is risk of unsafe memory access.
+        uint16_t metadataStartIndex = static_cast<uint16_t>(bufReader.OctetsRead() + mFileDesLength);
+        mMetadata                   = &bufStart[metadataStartIndex];
+        mMetadataLength             = static_cast<uint16_t>(aBuffer.DataLength() - metadataStartIndex);
+    }
+
+exit:
+    if (bufReader.StatusCode() != CHIP_NO_ERROR)
+    {
+        err = bufReader.StatusCode();
+    }
+    return err;
+}
+
+size_t TransferInit::PackedSize() const
+{
+    bool widerange    = ((mStartOffset | mMaxLength) > std::numeric_limits<uint32_t>::max());
```

Comment:
this is more brittle than I'd like

instead of calculating here to match Pack(), you could run Pack() with a null buffer and return what BufBound tells you.  this too, is part of the intended use of BufBound's semantics

Reply:
So would this require having Pack() take a BufBound as an argument instead of a PacketBuffer? That's ok with me. It would also require that Pack() not make any checks for available space (and exit early before attempting to write everything), but I guess that's the whole point of BufBound.

I agree that the current implementation is brittle